src/katgpucbf/xbgpu/engine.py

```diff
@@ -259,9 +273,171 @@ def capture_enable(self, enable: bool = True) -> None:
         raise NotImplementedError  # pragma: nocover
 
 
+class BPipeline(Pipeline):
+    """Processing pipeline for a collection of :class:`~output.BOutput`."""
+
+    output: BOutput
+
+    def __init__(
+        self,
+        output: BOutput,
+        engine: "XBEngine",
+        context: AbstractContext,
+        init_tx_enabled: bool,
+        name: str = DEFAULT_BPIPELINE_NAME,
+    ) -> None:
+        super().__init__(output, name, engine, context)
+        self.n_beams = 1
+
+        # TODO: Obtain this in a neater way once the beamformer OpSequence is done
+        buffer_shape = (
+            engine.heaps_per_fengine_per_chunk,
+            self.n_beams,
+            engine.n_channels_per_substream,
+            engine.src_layout.n_spectra_per_heap,
+            COMPLEX,
+        )
+        for _ in range(self.n_tx_items):
+            buffer_device = accel.DeviceArray(context, shape=buffer_shape, dtype=bsend.SEND_DTYPE)
+            # TODO: Clarify shape of `saturated`
+            saturated = accel.DeviceArray(
+                context, shape=(engine.heaps_per_fengine_per_chunk, self.n_beams), dtype=np.uint32
+            )
+            present_ants = np.zeros(shape=(engine.n_ants,), dtype=bool)
+            tx_item = TxQueueItem(buffer_device, saturated, present_ants)
+            self._tx_free_item_queue.put_nowait(tx_item)
+
+        # TODO: The way this is imported will change once the OperationSequence is in place
+        self.send_stream = bsend.BSend(
+            output,
+            engine.heaps_per_fengine_per_chunk,
+            self.n_tx_items,
+            n_channels_per_substream=engine.n_channels_per_substream,
+            spectra_per_heap=engine.src_layout.n_spectra_per_heap,
+            timestamp_step=engine.rx_heap_timestamp_step,
+            send_rate_factor=engine.send_rate_factor,
+            channel_offset=engine.channel_offset_value,
+            context=context,
+            packet_payload=engine.dst_packet_payload,
```

Comment:
Hmm, it occurs to me that while we need the payload size to be same for all beams, there is no fundamental reason it has to be the same between X and B. It's not something to worry about too much now, but maybe file a ticket.

Reply:
I've created [NGC-1144](https://skaafrica.atlassian.net/browse/NGC-1144). Thanks for flagging it.